Remove PoC and revamp Samples (#412)
open-telemetry/opentelemetry-dotnet-instrumentation: samples/README.md

> # Samples

Reviewer: I'd more specific about what the samples are about in the heading.

Author: Kind of long, I'm open to other suggestions:
```suggestion
# Instrumenting Sample Applications with the OpenTelemetry .NET Auto-Instrumentation
```

> The samples are intended to demonstrate how different scenarios are covered by
> the OpenTelemetry .NET Auto-Instrumentation.

Reviewer: ```suggestion
The following samples show how different instrumentation scenarios are covered by
the OpenTelemetry .NET Auto-Instrumentation.
```

## Overview

> The sample script will instrument two .NET processes, these process are identified
> as two [OpenTelemetry services](https://github.com/open-telemetry/opentelemetry-specification/blob/d6bcc0cb072d8d6f6ced856f1f23c451648a3caa/specification/resource/semantic_conventions/README.md#service)
> with names:

Reviewer: ```suggestion
The sample script instruments two .NET processes. These processes are identified
as [OpenTelemetry services](https://github.com/open-telemetry/opentelemetry-specification/blob/d6bcc0cb072d8d6f6ced856f1f23c451648a3caa/specification/resource/semantic_conventions/README.md#service)
with the following names:
```

>  1. **`aspnet-server`**: an ASP.NET Core application that makes requests to MongoDB and Redis databases;
>  2. **`http-client`**: a .NET application that makes a few HTTP requests to the `aspnet-server` and some public websites;

Reviewer: ```suggestion
 1. **`aspnet-server`**: ASP.NET Core application that makes requests to MongoDB and Redis databases.
 2. **`http-client`**: .NET application that makes a few HTTP requests to the `aspnet-server` and some public websites.
```

> The script creates docker containers for MongoDB, Redis, and Jaeger.
> You can see the traces generated by the sample at the Jaeger container.
> The number and organization of traces generated depends on the `http-client`
> used in the specific run.
> The script will wait for your input before stopping the docker containers.

Reviewer: ```suggestion
The script creates Docker containers for MongoDB, Redis, and Jaeger.
You can see the traces generated by the sample in the Jaeger container.
The number and organization of traces generated depends on the `http-client`
used in the specific run. The script waits for your input before stopping the containers.
```

## How to Run the Samples

>  1. Create a local build of the project per [developing instructions](../docs/developing.md)
>  2. On the root of the repository run the `./run_sample.sh`

Reviewer: ```suggestion
 1. Create a local build of the project following the [developing instructions](../docs/developing.md).
 2. At the root of the repository, run the `./run_sample.sh` script.
```

> The script run can be customized via environment variables:

Reviewer: ```suggestion
The script run can be customized using the following environment variables:
```

| Environment variable | Description | Default |
|-|-|-|
| **aspNetAppTargetFramework** | Target framework for the `aspnet-server` service | `netcoreapp3.1` |
| **exporter** | Value for [`OTEL_TRACES_EXPORTER`](../docs/config.md#exporters) env. variable | `jaeger` |
| **sampleApp** | Application selected for the `http-client` service | `ConsoleApp` |

> | **sampleAppInjectSDK** | Boolean to control if Auto-instrumentation will inject the SDK on the `http-client` service | `true` |

Reviewer: ```suggestion
| **sampleAppInjectSDK** | Whether auto-instrumentation injects the SDK on the `http-client` service | `true` |
```

| **sampleAppTargetFramework** | Target framework for the `http-client` service | `netcoreapp3.1` |

Usage example:
```terminal

> $ sampleAppTargetFramework="net6.0" ./run-sample.sh

Author: I need to apply the same at other locations...

```

## Sample Projects

> Description of each project under the `samples` folder.

Reviewer: ```suggestion
You can find a description of each project in the `samples` folder.
```

Author: This gets a bit recursive since the file with the description is this one. 🤔 

```suggestion
You can find a description of each sample project below.
```

### BindingRedirect Project

> Sample client app that uses .NET Framework binding redirects to handle assembly version conflict
> between auto-instrumentation and 3rd party libraries (see the OldReference project)
> that can't be updated to the same version required by OpenTelemetry .NET SDK used by auto-instrumentation.

Reviewer: ```suggestion
Sample client app that uses .NET Framework binding redirects to handle assembly version conflict
between auto-instrumentation and third-party libraries that can't be updated to the same version required by OpenTelemetry. See the `OldReference` project for more information.
```

> The project representing the 3rd party library has a legacy Activity named `InstrumentedHttpCall.GetAsync`
> set [`OTEL_DOTNET_AUTO_LEGACY_SOURCES`](../docs/config.md#customization) to capture it.

Reviewer: ```suggestion
The project representing the third-party library has a legacy Activity named `InstrumentedHttpCall.GetAsync`
set [`OTEL_DOTNET_AUTO_LEGACY_SOURCES`](../docs/config.md#customization) to capture it.
```

> Usage example (for Windows only since binding redirection only applies to .NET Framework):

Reviewer: ```suggestion
The following usage example is for Windows, as binding redirection only applies to .NET Framework:
```

```terminal
OTEL_DOTNET_AUTO_LEGACY_SOURCES="InstrumentedHttpCall.GetAsync" sampleApp=BindingRedirect sampleAppTargetFramework=net472 ./run-sample.sh
```

### ConsoleApp

> The default sample app used by `./run-samples.sh`. It contains manual instrumentation using a
> [ActivitySource](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/README.md#activity-source)
> and [OpenTracing](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Shims.OpenTracing#readme).
> No configuration is needed to capture the manually created ActivitySource since its name is
> `OpenTelemetry.AutoInstrumentation.ConsoleApp` which is match for the default value for the environment variable
> [`OTEL_DOTNET_AUTO_ADDITIONAL_SOURCES`](../docs/config.md#customization).

Reviewer: ```suggestion
The default sample app used by `./run-samples.sh`. It contains manual instrumentation using an
[ActivitySource](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/README.md#activity-source)
and [OpenTracing](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Shims.OpenTracing#readme).
No configuration is needed to capture the manually created ActivitySource, as its name is
`OpenTelemetry.AutoInstrumentation.ConsoleApp`, which is match for the default value for the environment variable
[`OTEL_DOTNET_AUTO_ADDITIONAL_SOURCES`](../docs/config.md#customization).
```

Usage example:
```terminal
$ sampleAppTargetFramework="net6.0" ./run-sample.sh
```

### ConsoleApp.SelfBootstrap

A sample client app that takes care of bootstrapping the OpenTelemetry SDK. It uses an
[ActivitySource](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/README.md#activity-source)
to implement manual instrumentation. The bootstrap code takes care of adding the ActivitySource name
for the auto-instrumentation and the manual instrumentation.

The bootstrap code sets the [Zipkin exporter](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Zipkin/README.md)
provided by the SDK. The Jaeger container is also configured to accept Zipkin spans.

Use the script environment variable `sampleAppInjectSDK` to control the value of
[`OTEL_DOTNET_AUTO_LOAD_AT_STARTUP`](../docs/config.md#customization) used to launch
the sample app.

Usage example:
```terminal
$ sampleAppInjectSDK=false sampleApp=ConsoleApp.SelfBootstrap ./run-sample.sh
```

### CoreAppOldReference

> Sample client app that uses NuGet package version resolution at build time to handle assembly version conflict
> between auto-instrumentation and 3rd party libraries (see the OldReference project)
> that can't be updated to the same version required by OpenTelemetry .NET SDK used by auto-instrumentation.

Reviewer: ```suggestion
Sample client app that uses NuGet package version resolution at build time to handle assembly version conflict
between auto-instrumentation and third-party libraries that can't be updated to the same version required by OpenTelemetry. See the `OldReference` project for more information.
```

> The project representing the 3rd party library has a legacy Activity named `InstrumentedHttpCall.GetAsync`

Reviewer: ```suggestion
The project representing the third-party library has a legacy Activity named `InstrumentedHttpCall.GetAsync`
```

set [`OTEL_DOTNET_AUTO_LEGACY_SOURCES`](../docs/config.md#customization) to capture it.

Usage example:
```terminal
$ OTEL_DOTNET_AUTO_LEGACY_SOURCES="InstrumentedHttpCall.GetAsync" sampleApp=CoreAppOldReference ./run-sample.sh
```

### OldReference

> Represents a 3rd party project that can't be updated to use the same version of the assemblies
> required by OpenTelemetry .NET SDK used by auto-instrumentation.

Reviewer: ```suggestion
Represents a third-party project that can't be updated to use the same version of the assemblies
required by OpenTelemetry .NET SDK used by auto-instrumentation.
```

### OpenTracing.Library

A library that wraps an OpenTracing span around a `Func<Task>` delegate.

### Samples.AspNetCoreMvc31

> An ASP.NET Core MVC application used to simulate a server application capable of making calls
> to MongoDB and Redis. Besides these calls it also supports various other endpoints, check
> the [Controllers](./Samples.AspNetCoreMvc31/Controllers/) for some of the endpoints that
> can be used.

Reviewer: ```suggestion
An ASP.NET Core MVC application used to simulate a server application capable of making calls
to MongoDB and Redis. In addition to these calls it also supports various other endpoints. Check
the [Controllers](./Samples.AspNetCoreMvc31/Controllers/) for some of the endpoints that
can be used.
```
